pkg/export/disk/disk.go

```diff
@@ -0,0 +1,271 @@
+	}
+
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("error marshaling data: %w", err)
+	}
+
+	if conn.File == nil {
+		return fmt.Errorf("no file to write the violation in")
+	}
+
+	_, err = conn.File.WriteString(string(jsonData) + "\n")
+	if err != nil {
+		return fmt.Errorf("error writing message to disk: %w", err)
+	}
+
+	if violation.Message == "audit is completed" {
+		err := conn.handleAuditEnd(topic)
+		if err != nil {
+			return fmt.Errorf("error handling audit end: %w", err)
+		}
+		conn.File = nil
+		conn.currentAuditRun = ""
+		r.openConnections[connectionName] = conn
+	}
+	return nil
+}
+
+func (conn *Connection) handleAuditStart(auditID string, topic string) error {
+	conn.currentAuditRun = strings.ReplaceAll(auditID, ":", "_")
+
+	// Ensure the directory exists
+	dir := path.Join(conn.Path, topic)
+	if err := os.MkdirAll(dir, 0o755); err != nil {
+		return fmt.Errorf("failed to create directories: %w", err)
+	}
+
+	file, err := os.OpenFile(path.Join(dir, appendExtension(conn.currentAuditRun, "txt")), os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o644)
+	if err != nil {
+		return fmt.Errorf("failed to open file: %w", err)
+	}
+	conn.File = file
+	err = retry.OnError(retry.DefaultBackoff, func(_ error) bool {
+		return true
+	}, func() error {
+		return syscall.Flock(int(conn.File.Fd()), syscall.LOCK_EX)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to acquire lock: %w", err)
+	}
+	return nil
+}
+
+func (conn *Connection) handleAuditEnd(topic string) error {
+	if err := retry.OnError(retry.DefaultBackoff, func(_ error) bool {
+		return true
+	}, conn.unlockAndCloseFile); err != nil {
+		return fmt.Errorf("error closing file: %w, %s", err, conn.currentAuditRun)
+	}
+	conn.File = nil
+
+	readyFilePath := path.Join(conn.Path, topic, appendExtension(conn.currentAuditRun, "log"))
+	if err := os.Rename(path.Join(conn.Path, topic, appendExtension(conn.currentAuditRun, "txt")), readyFilePath); err != nil {
+		return fmt.Errorf("failed to rename file: %w, %s", err, conn.currentAuditRun)
+	}
+
+	return conn.cleanupOldAuditFiles(topic)
+}
+
+func (conn *Connection) unlockAndCloseFile() error {
+	if conn.File == nil {
+		return fmt.Errorf("no file to close")
+	}
+	fd := int(conn.File.Fd())
+	if fd < 0 {
+		return fmt.Errorf("invalid file descriptor")
+	}
+	if err := syscall.Flock(fd, syscall.LOCK_UN); err != nil {
+		return fmt.Errorf("failed to release lock: %w", err)
+	}
+	if err := conn.File.Close(); err != nil {
+		return fmt.Errorf("failed to close file: %w", err)
+	}
+	return nil
+}
+
+func (conn *Connection) cleanupOldAuditFiles(topic string) error {
+	dirPath := path.Join(conn.Path, topic)
+
+	for {
+		earliestFile, files, err := getEarliestFile(dirPath)
+		if err != nil {
+			return fmt.Errorf("error getting earliest file: %w", err)
+		}
+		if len(files) <= conn.MaxAuditResults {
+			break
+		}
+		if err := os.Remove(earliestFile); err != nil {
+			return fmt.Errorf("error removing file: %w", err)
+		}
```

Comment:
Noting that if the reader is still stuck processing an earlier file this will truncate the file since it doesn't try to acquire a lock first. Maybe that's okay or preferred to ensure we don't reach any deadlocks.